mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxAnnotationDAO.java

```diff
@@ -0,0 +1,141 @@
+import static org.apache.james.mailbox.postgres.PostgresMailboxAnnotationModule.PostgresMailboxAnnotationTable.TABLE_NAME;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.mailbox.model.MailboxAnnotation;
+import org.apache.james.mailbox.model.MailboxAnnotationKey;
+import org.apache.james.mailbox.postgres.PostgresMailboxId;
+import org.jooq.impl.DSL;
+import org.jooq.impl.DefaultDataType;
+import org.jooq.postgres.extensions.types.Hstore;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresMailboxAnnotationDAO {
+    private static final char SQL_WILDCARD_CHAR = '%';
+    private static final String ANNOTATION_KEY_FIELD_NAME = "annotation_key";
+    private static final String ANNOTATION_VALUE_FIELD_NAME = "annotation_value";
+
+    private final PostgresExecutor postgresExecutor;
+
+    public PostgresMailboxAnnotationDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    public Flux<MailboxAnnotation> getAllAnnotations(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext ->
+                Flux.from(dslContext.selectFrom(TABLE_NAME)
+                    .where(MAILBOX_ID.eq(mailboxId.asUuid()))))
+            .singleOrEmpty()
+            .map(record -> record.get(ANNOTATIONS, LinkedHashMap.class))
+            .flatMapIterable(this::hstoreToAnnotations);
+    }
+
+    public Flux<MailboxAnnotation> getAnnotationsByKeys(PostgresMailboxId mailboxId, Set<MailboxAnnotationKey> keys) {
+        return postgresExecutor.executeRows(dslContext ->
+                Flux.from(dslContext.select(DSL.function("slice",
+                        DefaultDataType.getDefaultDataType("hstore"),
+                        ANNOTATIONS,
+                        DSL.array(keys.stream().map(mailboxAnnotationKey -> DSL.val(mailboxAnnotationKey.asString())).collect(Collectors.toUnmodifiableList()))))
+                    .from(TABLE_NAME)
+                    .where(MAILBOX_ID.eq(mailboxId.asUuid()))))
+            .singleOrEmpty()
+            .map(record -> record.get(0, LinkedHashMap.class))
+            .flatMapIterable(this::hstoreToAnnotations);
+    }
+
+    public Mono<MailboxAnnotation> getAnnotationsByKey(PostgresMailboxId mailboxId, MailboxAnnotationKey key) {
+        return postgresExecutor.executeRows(dslContext ->
+                Flux.from(dslContext.select(DSL.field(ANNOTATIONS.getName() + " -> ?", key.asString()))
+                    .from(TABLE_NAME)
+                    .where(MAILBOX_ID.eq(mailboxId.asUuid()))))
+            .singleOrEmpty()
+            .map(record -> MailboxAnnotation.newInstance(key, record.get(0, String.class)));
```

Comment:
if `record.get(0, String.class)` return null, it will throw NullPointerException

I caught it when tried to run PostgresAnotationmapper for imap mpt test

```
java.lang.NullPointerException: null
	at java.base/java.util.Objects.requireNonNull(Objects.java:221)
	at java.base/java.util.Optional.<init>(Optional.java:107)
	at java.base/java.util.Optional.of(Optional.java:120)
	at org.apache.james.mailbox.model.MailboxAnnotation.newInstance(MailboxAnnotation.java:34)
	at org.apache.james.mailbox.postgres.mail.dao.PostgresMailboxAnnotationDAO.lambda$getAnnotationsByKey$6(PostgresMailboxAnnotationDAO.java:84)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:106)
	at reactor.core.publisher.Operators$MonoInnerProducerBase.complete(Operators.java:2811)
	at reactor.core.publisher.MonoSingle$SingleSubscriber.onComplete(MonoSingle.java:180)
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyInner.onComplete(MonoFlatMapMany.java:260)
	at org.jooq.impl.R2DBC$AbstractSubscription.lambda$new$0(R2DBC.java:172)
```

Reply:
fixed